aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1479 @@
+        synchronize_delegation_pool(pool);
+
+        account::create_signer_with_capability(&pool.stake_pool_signer_cap)
+    }
+
+    /// Get the address of delegation pool reference `pool`.
+    fun get_pool_address(pool: &DelegationPool): address {
+        account::get_signer_capability_address(&pool.stake_pool_signer_cap)
+    }
+
+    /// Return the index of the current lockup cycle on delegation pool `pool_address`.
+    /// This represents the key into `inactive_shares` of the `pending_inactive` shares pool.
+    public fun current_lockup_epoch(pool_address: address): u64 acquires DelegationPool {
+        assert_delegation_pool_exists(pool_address);
+        borrow_global<DelegationPool>(pool_address).current_lockup_epoch
+    }
+
+    /// Allows an owner to change the operator of the underlying stake pool.
+    public entry fun set_operator(owner: &signer, new_operator: address) acquires DelegationPoolOwnership, DelegationPool {
+        stake::set_operator(
+            &retrieve_stake_pool_signer(get_owned_pool_address(signer::address_of(owner))),
+            new_operator
+        );
+    }
+
+    /// Allows an owner to change the delegated voter of the underlying stake pool.
+    public entry fun set_delegated_voter(owner: &signer, new_voter: address) acquires DelegationPoolOwnership, DelegationPool {
+        stake::set_delegated_voter(
+            &retrieve_stake_pool_signer(get_owned_pool_address(signer::address_of(owner))),
+            new_voter
+        );
+    }
+
+    /// Add `amount` of coins to the delegation pool `pool_address`.
+    public entry fun add_stake(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // short-circuit if amount to add is 0 so no event is emitted
+        if (amount == 0) { return };
+
+        let stake_pool_signer = retrieve_stake_pool_signer(pool_address);
+        let pool = borrow_global_mut<DelegationPool>(pool_address);
+        let delegator_address = signer::address_of(delegator);
+
+        // extract from absolute added stake the maximum amount the delegator would unfairly earn this epoch
+        let amount_added = charge_add_stake_fee(pool, amount);
+
+        coin::transfer<AptosCoin>(delegator, signer::address_of(&stake_pool_signer), amount);
+        stake::add_stake(&stake_pool_signer, amount);
+        buy_in_active_shares(pool, delegator_address, amount_added);
+
+        event::emit_event(
+            &mut pool.add_stake_events,
+            AddStakeEvent {
+                pool_address,
+                delegator_address,
+                amount_added,
+            },
+        );
+    }
+
+    /// Unlock `amount` from the active + pending_active stake of `delegator` or
+    /// at most how much active stake there is on the stake pool.
+    public entry fun unlock(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // execute pending withdrawal if existing before creating a new one
```

Comment:
Can we validate that amount must be > 0? Let's protect the pending_withdrawal as much as possible and not update it with 0s if possible.